libraries/AP_HAL/CAN.h

```diff
@@ -0,0 +1,233 @@
+                    1 - CAN port is initialized
+     */
+    virtual bool is_initialized() = 0;
+
+    /*
+        Return CAN bus type for opened port
+
+        return      CANBUS_TYPE_UNKNOWN - if CAN port not opened or initialized
+                    One of CANBUS_TYPE_ - if CAN port is opened and initialized
+     */
+    virtual CanBusType get_bus_type() = 0;
+
+    /*
+        Return if CAN port has some untransmitted pending messages
+
+        return     -1 - CAN port is not opened or initialized
+                    0 - no messages are pending
+             positive - number of pending messages
+     */
+    virtual int32_t tx_pending() = 0;
+
+    /*
+        Return if CAN port has received but yet read messages
+
+        return     -1 - CAN port is not opened or initialized
+                    0 - no messages are pending for read
+             positive - number of pending messages for read
+     */
+    virtual uint32_t available() = 0;
+
+    /*
+        Read pending message
+
+        ident       Reference to a variable that will contain identifier
+                    11 or 29 bit depending on can bus type
+        buffer      Pointer to data buffer. Payload will be written to this buffer.
+        size        Number of bytes in payload that was read
+
+        return      0 - Message is not read. If due to error then code of error
+                        can be read with get_last_error() method
+                    1 - Message is successfully read
+     */
+    virtual bool read(uint32_t &ident, uint8_t *buffer, uint8_t &size) = 0;
+
+    /*
+        Write message
+
+        ident       Identifier of message to be transmitted
+                    11 or 29 bit depending on can bus type
+        buffer      Pointer to data buffer containing payload
+        size        Number of bytes in payload to be transmitted
+
+        return      0 - Message is not written for sending. If due to error then code of error
+                        can be read with get_last_error() method
+                    1 - Message is successfully written for sending
+     */
+    virtual bool write(uint32_t ident, const uint8_t *buffer, uint8_t size) = 0;
+
+    /*
+        Get system timestamp of last read message with read(uint32_t &ident, uint8_t *buffer, uint8_t &size)
+        Returns the system time when the message was actually received by hardware.
+        This method might not be available on all HW platforms.
+
+        return      System timestamp value
+     */
+    virtual uint32_t get_last_read_timestamp() = 0;
+
+    /*
+        Get system timestamp of last written message with write(uint32_t ident, const uint8_t *buffer, uint8_t size)
+        Returns the system time when the message was actually written by hardware.
+        This method might not be available on all HW platforms.
+
+        return      System timestamp value
+     */
+    virtual uint32_t get_last_write_timestamp() = 0;
```

Comment:
All messages do not need to be timestamped. Maybe add additional parameter to write method that will save timestamp to be read with get_last_write_timestamp? That way we will know for sure that it is a timestamp for synchro message as it is not send out very often.